### PR TITLE
Get started with Beta.14 upgrade guide

### PR DESCRIPTION
**Pull request by franzliedke**
Let's get this ready so that extension developers can prepare themselves for Beta.14.

Since this will be a relatively complex upgrade for those not familiar with the internals of Mithril and its different versions, let's go above and beyond in helping them understand what needs to be done.

Refs flarum/core#2201.

### Open TODOs:

- [ ] Finish section on backend changes
- [ ] Mention removal of auth-* extensions, see flarum/core#2006

### CODE REVIEW

File: docs/extend/update-b14.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+When you have taken care of the changes above, you should be good to go.
+For the following changes, we have bought you time until the stable release.
+Considering you have to do the changes anyway, why not do them now?
+
+*TODO*
+
+- `this.props` -> `this.attrs`
+- static `initProps()` -> static `initAttrs()`
+- `m.prop` -> `m.stream`
+- `m.withAttr` -> `withAttr` with import
```
Comment (franzliedke) [inline docs/extend/update-b14.md:75]: This is my proposal for a minimal BC layer. These... 

a) are trivial to add and maintain,
b) will be removed for the stable release, and
c) are repetitive and frequent, so avoiding them now should help the extension ecosystem.

File: docs/extend/update-b14.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  - register routes with page class, not instance
+    - special case when passing props
+  - `<a href={url} config={m.route}>` -> `<a route={url}>`
+- AJAX
+  - `m.request({...})` -> `data:` key split up into `body:` and `params:`
+  - `m.deferred` -> native `Promise`
+- Redrawing
+  - `m.redraw(true)` -> `m.redraw.sync()`
+  - `m.redraw.strategy('none')` -> `e.redraw = false` in event handler
+  - `m.lazyRedraw()` -> `m.redraw()`
```
Comment (franzliedke) [inline docs/extend/update-b14.md:59]: These will need to be explained thoroughly. But for now I wanted to collect the steps here. Is there anything missing (here or in the `Optional` section below) related to the Mithril 2.0 upgrade?

**Comment by datitisev** [comment]
I think we should mention that having empty strings as Element attributes will now error - probably not many people do this but I had some code and it took me a while until I was able to locate the problem.

Some code like this (looks ugly but I'd rather have it inline LOL)

image

`''` can be replaced with `null` and it'll work fine.

File: docs/extend/update-b14.md (hunk 1 above)
Comment (datitisev) [reply docs/extend/update-b14.md:75]: Not sure what you mean with this comment? Do you want to remove all of these?

File: docs/extend/update-b14.md (hunk 2 above)
Comment (datitisev) [reply docs/extend/update-b14.md:59]: I'd add that you can no longer use a `children` attribute unless the component extracts it before passing it as like `{...attrs}` or something.

File: docs/extend/update-b14.md [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+
+```js
+class ChildComponent extends Component {
+  view() {
+    return <p>{this.attrs.counter}</p>;
+  }
+}
+
+class ParentComponent extends Component {
+  init() {
```
Comment (datitisev) [inline docs/extend/update-b14.md:173]: This should be changed to `oninit` and call `super.oninit(...)`

**Review by KyrneDev — COMMENTED** [review]
Looks good in current state. Only issue I see is there are a few comments before JS code blocks that are annotated as mithril 0.2 when it's actually 2.0

**Review by askvortsov1 — APPROVED** [review]

**Pull request merged**